### PR TITLE
Factory Method Example

### PR DESCRIPTION
**Pull request by RohollahHosseini**

### CODE REVIEW

File: Examples/MyExample(FactoryMethod)/TransportService.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            Console.WriteLine("Booking via TrainTransport.");
+        }
+    }
+
+    public interface ITransportFactory
+    {
+        ITransport CreateTransport();
+    }
+
+    public class AirTransportationServiceFactory : ITransportFactory
```
Comment (mohammadKarimi) [inline Examples/MyExample(FactoryMethod)/TransportService.cs:70]: you have created factory for Air. so why did you hold your instance management in your concrete class? 
it is factory and it's responsibility is to create and manage the instance creation.

File: Examples/MyExample(FactoryMethod)/TransportService.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        private static readonly object _lock = new();
+
+        private AirTransportationService() { }
+
+        public static AirTransportationService GetInstance()
+        {
+            lock (_lock)
+            {
+                if (_airTransportationService == null)
+                    _airTransportationService = new AirTransportationService();
```
Comment (mohammadKarimi) [inline Examples/MyExample(FactoryMethod)/TransportService.cs:21]: it is not a good idea to mange the instance creation here. you have to move it to related factory.
SOC Principle.

File: Examples/MyExample(FactoryMethod)/ProgramExample.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿namespace Examples.MyExample_FactoryMethod_;
+
+public class ProgramExample
+{
+    public static void RunApplication()
+    {
+
+        ITransportFactory airFactory = new AirTransportationServiceFactory();
```
Comment (mohammadKarimi) [inline Examples/MyExample(FactoryMethod)/ProgramExample.cs:8]: it is not a good idea to connect your user with your factory in straightforward.
for this case it is enough.

you can follow the diagram of factory method for more information.

**Review by mohammadKarimi — COMMENTED** [review]
now move to E3.1

Imagine it is a library and we make it as a package. so users or customers, can't change your code, but they want to add a new behavior like Truck. so, how you can solve it.

**Comment by RohollahHosseini** [comment]
Mr, @mohammadKarimi 
Please review my new code

**Pull request closed without merging**